File: backend/src/BMPCommerce.Application/Operations/Produtos/ProdutoService.cs
using BMPCommerce.Application.Common.Abstractions;
using BMPCommerce.Application.Common.Exceptions;
using BMPCommerce.Domain.Common;
using BMPCommerce.Domain.Entities;
using BMPCommerce.Domain.Enums;
using BMPCommerce.Domain.Interfaces;

namespace BMPCommerce.Application.Operations.Produtos;

public class ProdutoService : IProdutoService
{
    private readonly IProdutoRepository _produtoRepository;
    private readonly IVendaRepository _vendaRepository;
    private readonly IApplicationDbContext _dbContext;

    public ProdutoService(
        IProdutoRepository produtoRepository,
        IVendaRepository vendaRepository,
        IApplicationDbContext dbContext)
    {
        _produtoRepository = produtoRepository;
        _vendaRepository = vendaRepository;
        _dbContext = dbContext;
    }

    public async Task<IReadOnlyList<ProdutoDto>> ListarAsync(string? search, CancellationToken cancellationToken)
    {
        var produtos = await _produtoRepository.GetAllAsync(search, cancellationToken);
        return produtos.Select(MapToDto).ToList();
    }

    public async Task<ProdutoDto> ObterPorIdAsync(Guid id, CancellationToken cancellationToken)
    {
        var produto = await _produtoRepository.GetByIdAsync(id, cancellationToken)
            ?? throw new NotFoundException("Produto não encontrado.");

        return MapToDto(produto);
    }

    public async Task<Result<ProdutoDto>> CriarAsync(CriarProdutoRequest request, CancellationToken cancellationToken)
    {
        if (await _produtoRepository.ExisteSkuAsync(request.Sku, ignorarId: null, cancellationToken))
        {
            return Result<ProdutoDto>.Failure("Já existe um produto com esse SKU.");
        }

        if (!Enum.TryParse<UnidadeMedida>(request.UnidadeMedida, out var unidadeMedida))
        {
            return Result<ProdutoDto>.Failure("Unidade de medida inválida.");
        }

        var produto = new Produto(
            request.Nome,
            request.Sku,
            request.Descricao,
            request.CodigoBarras,
            request.Categoria,
            unidadeMedida,
            request.PrecoCusto,
            request.PrecoVenda,
            request.EstoqueAtual,
            request.EstoqueMinimo);

        _produtoRepository.Add(produto);
        await _dbContext.SaveChangesAsync(cancellationToken);

        return Result<ProdutoDto>.Success(MapToDto(produto));
    }

    public async Task<Result<ProdutoDto>> AtualizarAsync(Guid id, AtualizarProdutoRequest request, CancellationToken cancellationToken)
    {
        var produto = await _produtoRepository.GetByIdAsync(id, cancellationToken)
            ?? throw new NotFoundException("Produto não encontrado.");

        if (await _produtoRepository.ExisteSkuAsync(request.Sku, id, cancellationToken))
        {
            return Result<ProdutoDto>.Failure("Já existe um produto com esse SKU.");
        }

        if (!Enum.TryParse<UnidadeMedida>(request.UnidadeMedida, out var unidadeMedida))
        {
            return Result<ProdutoDto>.Failure("Unidade de medida inválida.");
        }

        produto.Atualizar(
            request.Nome,
            request.Sku,
            request.Descricao,
            request.CodigoBarras,
            request.Categoria,
            unidadeMedida,
            request.PrecoCusto,
            request.PrecoVenda,
            request.EstoqueAtual,
            request.EstoqueMinimo,
            request.Ativo);

        await _dbContext.SaveChangesAsync(cancellationToken);

        return Result<ProdutoDto>.Success(MapToDto(produto));
    }

    public async Task<Result> ExcluirAsync(Guid id, CancellationToken cancellationToken)
    {
        var produto = await _produtoRepository.GetByIdAsync(id, cancellationToken)
            ?? throw new NotFoundException("Produto não encontrado.");

        // Preserva o histórico de vendas (Doc 01 REGRA 4): produto vendido não sai do banco.
        if (await _vendaRepository.ExisteVendaComProdutoAsync(id, cancellationToken))
        {
            return Result.Failure("Produto possui vendas registradas e não pode ser excluído. Inative-o em vez de excluir.");
        }

        _produtoRepository.Remove(produto);
        await _dbContext.SaveChangesAsync(cancellationToken);

        return Result.Success();
    }

    private static ProdutoDto MapToDto(Produto produto) => new(
        produto.Id,
        produto.Nome,
        produto.Descricao,
        produto.Sku,
        produto.CodigoBarras,
        produto.Categoria,
        produto.UnidadeMedida.ToString(),
        produto.PrecoCusto,
        produto.PrecoVenda,
        produto.EstoqueAtual,
        produto.EstoqueMinimo,
        produto.Ativo,
        produto.CreatedAt,
        produto.UpdatedAt);
}

File: backend/app/services/produto_service.py
"""Equivalente a Application/Operations/Produtos/ProdutoService.cs."""

from __future__ import annotations

from decimal import Decimal
from uuid import UUID

from sqlalchemy.orm import Session

from app.core.exceptions import NotFoundException
from app.domain.common import Result
from app.domain.enums import UnidadeMedida
from app.domain.produto import Produto
from app.repositories.produto_repository import ProdutoRepository
from app.repositories.venda_repository import VendaRepository
from app.schemas.produto import AtualizarProdutoRequest, CriarProdutoRequest, ProdutoDto


class ProdutoService:
    def __init__(self, session: Session) -> None:
        self._session = session
        self._produtos = ProdutoRepository(session)
        self._vendas = VendaRepository(session)

    def listar(self, search: str | None) -> list[ProdutoDto]:
        produtos = self._produtos.get_all(search)
        return [self._to_dto(p) for p in produtos]

    def obter_por_id(self, produto_id: UUID) -> ProdutoDto:
        produto = self._produtos.get_by_id(produto_id)
        if produto is None:
            raise NotFoundException("Produto não encontrado.")

        return self._to_dto(produto)

    def criar(self, request: CriarProdutoRequest) -> Result[ProdutoDto]:
        if self._produtos.existe_sku(request.sku, ignorar_id=None):
            return Result.failure("Já existe um produto com esse SKU.")

        try:
            unidade_medida = UnidadeMedida(request.unidade_medida)
        except ValueError:
            return Result.failure("Unidade de medida inválida.")

        produto = Produto(
            nome=request.nome,
            sku=request.sku,
            descricao=request.descricao,
            codigo_barras=request.codigo_barras,
            categoria=request.categoria,
            unidade_medida=unidade_medida,
            preco_custo=_to_decimal(request.preco_custo),
            preco_venda=_to_decimal(request.preco_venda),
            estoque_atual=request.estoque_atual,
            estoque_minimo=request.estoque_minimo,
        )

        self._produtos.add(produto)
        self._session.commit()

        return Result.success(self._to_dto(produto))

    def atualizar(self, produto_id: UUID, request: AtualizarProdutoRequest) -> Result[ProdutoDto]:
        produto = self._produtos.get_by_id(produto_id)
        if produto is None:
            raise NotFoundException("Produto não encontrado.")

        if self._produtos.existe_sku(request.sku, ignorar_id=produto_id):
            return Result.failure("Já existe um produto com esse SKU.")

        try:
            unidade_medida = UnidadeMedida(request.unidade_medida)
        except ValueError:
            return Result.failure("Unidade de medida inválida.")

        produto.atualizar(
            nome=request.nome,
            sku=request.sku,
            descricao=request.descricao,
            codigo_barras=request.codigo_barras,
            categoria=request.categoria,
            unidade_medida=unidade_medida,
            preco_custo=_to_decimal(request.preco_custo),
            preco_venda=_to_decimal(request.preco_venda),
            estoque_atual=request.estoque_atual,
            estoque_minimo=request.estoque_minimo,
            ativo=request.ativo,
        )

        self._produtos.update(produto)
        self._session.commit()

        return Result.success(self._to_dto(produto))

    def excluir(self, produto_id: UUID) -> Result[None]:
        produto = self._produtos.get_by_id(produto_id)
        if produto is None:
            raise NotFoundException("Produto não encontrado.")

        # Preserva o histórico de vendas (Doc 01 REGRA 4): produto vendido não sai do banco.
        if self._vendas.existe_venda_com_produto(produto_id):
            return Result.failure(
                "Produto possui vendas registradas e não pode ser excluído. Inative-o em vez de excluir."
            )

        self._produtos.remove(produto)
        self._session.commit()

        return Result.success()

    @staticmethod
    def _to_dto(produto: Produto) -> ProdutoDto:
        return ProdutoDto.model_validate(produto)


def _to_decimal(value: float) -> Decimal:
    """`float` chega do schema Pydantic (contrato JSON com o frontend); convertido via
    `str()` para não herdar erro de ponto flutuante binário (Decimal(0.1) != Decimal("0.1"))."""
    return Decimal(str(value))
